File: python/108. Convert Sorted Array to Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def bst(start, end):
            if (start <= end):
                rootIndex = start + (end - start) // 2

                root = TreeNode(nums[rootIndex])
                root.left = bst(start, rootIndex - 1)
                root.right = bst(rootIndex + 1, end)

                return root

        return bst(0, len(nums) - 1)

File: java/108. Convert Sorted Array to Binary Search Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int[] nums;
    public TreeNode sortedArrayToBST(int[] nums) {
        this.nums = nums;

        return bst(0, nums.length - 1);

    }


    public TreeNode bst (int start, int end) {
        //base
        if (start <= end) {
            int rootIndex = start + (end - start) / 2;
            TreeNode root = new TreeNode(nums[rootIndex]);

            root.left = bst(start, rootIndex - 1);
            root.right = bst(rootIndex + 1, end);

            return root;
        }
        return null;
    }
}